### PR TITLE
Added extension for parameter adaptation during the training process

### PR DESCRIPTION
**Pull request by dmitriy-serdyuk**

### CODE REVIEW

**Comment by dwf** [comment]
This really needs a better name, I think.

**Comment by dmitriy-serdyuk** [comment]
Any suggestions? `ModifyParameter`? As I understand, one should use a verb as an extension name.

File: blocks/extensions/training.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+    Applies function to compute the new value of a shared parameter each
+    iteration.
+
+    This class can be used to adapt over the training process parameters like
+    learning rate, momentum, etc.
+
+    Parameters
+    ----------
+    parameter : shared variable to be adjusted
```
Comment (bartvm) [inline blocks/extensions/training.py]: Just a note about the docstring formatting, it's:
```
parameter_name : type
    A lengthy explanation of what this
    variable does
```
So in this case:
```
parameter : :class:`~tensor.TensorSharedVariable`
    The shared variable to be adjusted.
function : callable
    Function that takes the total number of examples seen (``int``)
    as an input, and outputs a numeric value to which the
    given shared variable will be set.
```

File: blocks/extensions/training.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    """
+    def __init__(self, parameter, function, **kwargs):
+        super(AdjustParameter, self).__init__(**kwargs)
+        self.learning_rate = parameter
+        self.function = function
+        self.num_examples = 0
+
+    def after_batch(self, batch):
+        self.num_examples += batch.values()[0].shape[0]
+        self.learning_rate.set_value(self.function(self.num_examples))
```
Comment (bartvm) [inline blocks/extensions/training.py]: Perhaps this function could also take the current value of `self.learning_rate` as an input? That we `function` can just be `def decay(current_value, num_examples_seen): current_value * 0.9` if you want to do something batch-wise instead of example-wise.

**Comment by bartvm** [comment]
Methods and functions should have verbs or verb phrases as names, but classes don't have to. So something like `SharedVariableAdjustor`, `SharedVariableUpdater`, `SharedVariableModifier`, etc. would also work.

File: blocks/extensions/training.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    Parameters
+    ----------
+    parameter : shared variable to be adjusted
+    function : function which input number of aggregated examples and outputs
+               the parameter value
+
+    """
+    def __init__(self, parameter, function, **kwargs):
+        super(AdjustParameter, self).__init__(**kwargs)
+        self.learning_rate = parameter
```
Comment (bartvm) [inline blocks/extensions/training.py]: If you want this class to be generic enough for updating any shared variable (which I think is a good idea), then this attribute should probably be `self.parameter` and not `self.learning_rate`.

File: blocks/extensions/training.py [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+        self.function = function
+        self.num_examples = 0
```
Comment (bartvm) [inline blocks/extensions/training.py]: There is no need to keep track of this yourself! The total number of examples seen so far is stored in the log. I'm not entirely sure about the precise attribute (@rizar?) but I think it should be something like `self.main_loop.log.current_row.iterations_done`.

File: blocks/extensions/training.py [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+    Parameters
+    ----------
+    parameter : shared variable to be adjusted
+    function : function which input number of aggregated examples and outputs
+               the parameter value
+
+    """
+    def __init__(self, parameter, function, **kwargs):
+        super(AdjustParameter, self).__init__(**kwargs)
```
Comment (bartvm) [inline blocks/extensions/training.py]: You should probably set a default callback here i.e. `kwargs.setdefault("after_every_batch", True)` will make sure this extension is run after each batch by default.

File: blocks/extensions/training.py [hunk 6]
```diff
@@ -0,0 +1,45 @@
+import inspect
+from blocks.extensions import TrainingExtension
+
+
+class SharedVariableModifier(TrainingExtension):
```
Comment (bartvm) [inline blocks/extensions/training.py]: You should probably inherit from `SimpleExtension` instead and move `after_batch` to a `do` method. The advantage is then that the user could choose to e.g. run this only once an epoch, or every `n` batches, etc. making it a bit more flexible (and fixing the current Travis error).

**Comment by bartvm** [comment]
This looks like a good step towards learning rate scheduling! I'll let @rizar have a quick look in case he had any other ideas on how to do this, but as far as I'm concerned this looks good.

File: blocks/extensions/training.py [hunk 7]
```diff
@@ -0,0 +1,45 @@
+import inspect
+from blocks.extensions import SimpleExtension
+
+
+class SharedVariableModifier(SimpleExtension):
+    """Adjusts shared variable parameter using some function.
+
+    Applies function to compute the new value of a shared parameter each
```
Comment (rizar) [inline blocks/extensions/training.py]: a function

File: blocks/extensions/training.py [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+    iteration.
+
+    This class can be used to adapt over the training process parameters
+    like learning rate, momentum, etc.
+
+    Parameters
+    ----------
+    parameter : :class:`~tensor.TensorSharedVariable`
+        shared variable to be adjusted
```
Comment (rizar) [inline blocks/extensions/training.py]: Sorry for being picky, but we everywhere else in Blocks the parameter docstrings are started with a capital letter.

**Comment by rizar** [comment]
LGTM after fixing docs. If you want to do it ideally, use the `unittest` package to join two tests you added  and reduce code duplication. But to be honest, we have copy-pasted tests everywhere in Blocks and this much better than not to have them at all.
